src/Features/Core/Portable/ExtractMethod/MethodExtractor.cs

```diff
@@ -60,43 +64,60 @@ public async Task<ExtractMethodResult> ExtractMethodAsync(CancellationToken canc
 
             operationStatus = CheckVariableTypes(analyzeResult.Status.With(operationStatus), analyzeResult, cancellationToken);
             if (operationStatus.Failed())
-                return new FailedExtractMethodResult(operationStatus);
+                return ExtractMethodResult.Fail(operationStatus);
 
             var insertionPointNode = GetInsertionPointNode(analyzeResult, cancellationToken);
 
             if (!CanAddTo(originalSemanticDocument.Document, insertionPointNode, out var canAddStatus))
-                return new FailedExtractMethodResult(canAddStatus);
+                return ExtractMethodResult.Fail(canAddStatus);
 
             cancellationToken.ThrowIfCancellationRequested();
+            var codeGenerator = this.CreateCodeGenerator(analyzeResult);
 
-            var (analyzedDocument, insertionPoint) = await GetAnnotatedDocumentAndInsertionPointAsync(
-                originalSemanticDocument, analyzeResult, insertionPointNode, cancellationToken).ConfigureAwait(false);
+            var statements = codeGenerator.GetNewMethodStatements(insertionPointNode, cancellationToken);
+            if (statements.Status.Failed())
+                return ExtractMethodResult.Fail(statements.Status);
 
-            var triviaResult = await PreserveTriviaAsync(OriginalSelectionResult.With(analyzedDocument), cancellationToken).ConfigureAwait(false);
-            cancellationToken.ThrowIfCancellationRequested();
+            return ExtractMethodResult.Success(
+                operationStatus,
+                async cancellationToken =>
```

Comment:
this is the crux of hte change.  first, we do frontload some statement checks, since tehy are required to know if we're showing the EM option or not.  but everything after that moves into this lazy computation in the success object.